pkg/crunch42/crunch.go

```diff
@@ -0,0 +1,226 @@
+// SOFTWARE.
+package crunch
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+
+	"moul.io/http2curl"
+)
+
+const baseURL = "https://platform.42crunch.com/api/v1"
+
+const debug = "CRUNCH_DEBUG"
+
+func (c *Client) GetCollection(id string) (*Item, *ErrorResponse, error) {
+
+	collection := &Item{}
+
+	path := fmt.Sprintf("%s/collections/%s", c.baseURL, id)
+	resp, err := c.DoRequest("GET", path, nil, collection)
+	if err != nil {
+		return nil, resp, err
+
+	}
+	return collection, resp, err
+}
+
+func (c *Client) ListCollections() (*Collections, *ErrorResponse, error) {
+	collections := &Collections{}
+
+	path := fmt.Sprintf("%s/collections", c.baseURL)
+	resp, err := c.DoRequest("GET", path, nil, collections)
+	if err != nil {
+		return nil, resp, err
+
+	}
+	return collections, resp, err
+}
+
+func (c *Client) CreateCollection(collection *Collection) (*Item, *ErrorResponse, error) {
+
+	toReturn := &Item{}
+	path := fmt.Sprintf("%s/collections", c.baseURL)
+	resp, err := c.DoRequest("POST", path, collection, toReturn)
+	if err != nil {
+		return nil, resp, err
+
+	}
+	return toReturn, resp, err
+}
+
+func (c *Client) CreateAPI(api *API) (*Item, *ErrorResponse, error) {
+	i := &Item{}
+	path := "https://platform.42crunch.com/api/v2/apis"
+
+	resp, err := c.DoRequest("POST", path, api, i)
+	if err != nil {
+		return nil, resp, err
+
+	}
+	return i, resp, err
+}
+
+type Client struct {
+	client  *http.Client
+	apiKey  string
+	baseURL *url.URL
+}
+
+func NewClient(apiKey string, httpClient *http.Client) (*Client, error) {
+	if httpClient == nil {
+		httpClient = http.DefaultClient
+	}
+
+	url, err := url.Parse(baseURL)
+	if err != nil {
+		return nil, err
+	}
+	client := &Client{
+		apiKey:  apiKey,
+		baseURL: url,
+		client:  httpClient,
+	}
+
+	return client, nil
+}
+
+func (c *Client) DoRequest(method, path string, body, v interface{}) (*ErrorResponse, error) {
+	req, err := c.NewRequest(method, path, body)
+	if err != nil {
+		return nil, err
+	}
+
+	return c.Do(req, v)
+}
+
+func (c *Client) DoFormRequestPOST(method, path string, api *API, v interface{}) (*ErrorResponse, error) {
+
+	return nil, nil
+
+}
+
+func (c *Client) NewRequest(method, path string, body interface{}) (*http.Request, error) {
+	// relative path to append to the endpoint url, no leading slash please
+	if path[0] == '/' {
+		path = path[1:]
+	}
+	rel, err := url.Parse(path)
+	if err != nil {
+		return nil, err
+	}
+
+	u := c.baseURL.ResolveReference(rel)
+	var req *http.Request
+	if body != nil {
+		bodyBytes, _ := json.Marshal(body)
+		req, _ = http.NewRequest(method, u.String(), bytes.NewBuffer(bodyBytes))
+		req.Header.Set("Content-Type", "application/json")
+	} else {
+		req, _ = http.NewRequest(method, u.String(), nil)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	req.Close = true
+
+	if _, ok := os.LookupEnv(debug); ok {
+		command, _ := http2curl.GetCurlCommand(req)
```

Comment:
Please add `error` checking here.

Reply:
The method is just reading request https://github.com/moul/http2curl/blob/master/http2curl.go#L57 and it would fail here. No point in error check here as it will produce the same error in lines 134-145 i.e. If the request was able to execute then getting curl output will check out correctly

Reply:
OK.